graphql_compiler/tests/test_helpers.py

```diff
@@ -772,82 +775,91 @@ def get_sqlalchemy_schema_info(dialect: str = "mssql") -> SQLAlchemySchemaInfo:
             "from_table": "Animal",
             "to_table": "FeedingEvent",
-            "from_column": "fed_at",
-            "to_column": "uuid",
+            "column_pairs": {("fed_at", "uuid")},
         },
         {
             "name": "Animal_BornAt",
             "from_table": "Animal",
             "to_table": "BirthEvent",
-            "from_column": "born_at",
-            "to_column": "uuid",
+            "column_pairs": {
+                ("birth_uuid", "uuid"),
+                ("birth_date", "event_date"),
+            },
         },
         {
             "name": "Animal_LivesIn",
             "from_table": "Animal",
             "to_table": "Location",
-            "from_column": "lives_in",
-            "to_column": "uuid",
+            "column_pairs": {("lives_in", "uuid")},
         },
         {
             "name": "Animal_ImportantEvent",
             "from_table": "Animal",
             "to_table": "Union__BirthEvent__Event__FeedingEvent",
-            "from_column": "important_event",
-            "to_column": "uuid",
+            "column_pairs": {("important_event", "uuid")},
         },
         {
             "name": "Species_Eats",
             "from_table": "Species",
             "to_table": "Union__Food__FoodOrSpecies__Species",
-            "from_column": "eats",
-            "to_column": "uuid",
+            "column_pairs": {("eats", "uuid")},
         },
         {
             "name": "Entity_Related",
             "from_table": "Entity",
             "to_table": "Entity",
-            "from_column": "related_entity",
-            "to_column": "uuid",
+            "column_pairs": {("related_entity", "uuid")},
         },
         {
             "name": "Event_RelatedEvent",
             "from_table": "Union__BirthEvent__Event__FeedingEvent",
             "to_table": "Union__BirthEvent__Event__FeedingEvent",
-            "from_column": "related_event",
-            "to_column": "uuid",
+            "column_pairs": {("related_event", "uuid")},
         },
         {
             "name": "Entity_Alias",
             "from_table": "Entity",
             "to_table": "Alias",
-            "from_column": "uuid",
-            "to_column": "alias_for",
+            "column_pairs": {("uuid", "alias_for")},
         },
     ]
 
-    join_descriptors: Dict[str, Dict[str, DirectJoinDescriptor]] = {}
+    # Create the appropriate JoinDescriptor for each specified edge, in both the
+    # in and out directions.
+    join_descriptors: Dict[str, Dict[str, JoinDescriptor]] = {}
     for edge in edges:
-        join_descriptors.setdefault(edge["from_table"], {})[
-            "out_{}".format(edge["name"])
-        ] = DirectJoinDescriptor(edge["from_column"], edge["to_column"])
-        join_descriptors.setdefault(edge["to_table"], {})[
-            "in_{}".format(edge["name"])
-        ] = DirectJoinDescriptor(edge["to_column"], edge["from_column"])
+        column_pairs = cast(Set[Tuple[str, str]], edge["column_pairs"])
```

Comment:
This cast here is needed because the `edges` type is not specified and is probably inferred incorrectly. This is test code and is correct, so I don't have strong feelings one way or the other — just wanted to flag it in case you wanted to reconsider the alternative.

Reply:
Yeah there's no way to add an inline type hint that will fix this, and I don't want to blow up this PR